File: src/09-challenge/solution.py
from typing import List, Literal

type Move = Literal["U", "D", "R", "L"]
type Result = Literal["none", "crash", "eat"]


def move_train(board: List[str], mov: Move) -> Result:
    move_actions: dict[Move, int] = {
        "D": len(board[0]) + 1,
        "L": -1,
        "R": 1,
        "U": -len(board[0]) - 1,
    }

    result_actions: dict[str, Result] = {
        "·": "none",
        "*": "eat",
        "|": "crash",
        "o": "crash",
        "undefined": "crash",
    }

    board_stringified: str = "|".join(board)

    train_i: int = board_stringified.index("@")
    if train_i == -1:
        return "none"

    next_train_cell: str = board_stringified[train_i + move_actions[mov]]

    return result_actions[next_train_cell]

File: src/09-challenge/solution.js
/**
 * @param {string[]} board - Represent the train situation
 * @param {'U' | 'D' | 'R' | 'L'} mov - Movement direction
 * @returns {'none' | 'crash' | 'eat'}
 */
export default function moveTrain(board, mov) {
	/** @type {Object.<'U' | 'D' | 'R' | 'L', number>} */
	const moveActions = {
		D: board[0].length + 1,
		L: -1,
		R: 1,
		U: -board[0].length - 1,
	}

	/** @type {Object.<'·' | '*' | '|' | 'o' | 'undefined', 'none' | 'crash' | 'eat'>} */
	const resultActions = {
		'·': 'none',
		'*': 'eat',
		'|': 'crash',
		'o': 'crash',
		'undefined': 'crash',
	}

	/** @type {string} */
	const boardStringified = board.join('|')

	/** @type {number} */
	const trainI = boardStringified.indexOf('@')
	if (trainI === -1) return 'none'

	/** @type {string} */
	const nextTrainCell = boardStringified[trainI + moveActions[mov]]

	return resultActions[nextTrainCell]
}
